### PR TITLE
Put record parameters in scope withing instance initializers

### LINKED ISSUE #44879: It looks like copy constructor attempts to copy static fields
```
data class C(int x)
{
    static int y = 123;
}
```

Observed:
```
warning CS1717: Assignment made to same variable; did you mean to assign something else?
```

```
    .method public hidebysig specialname rtspecialname 
        instance void .ctor (
            class C ''
        ) cil managed 
    {
        // Method begins at RVA 0x20a8
        // Code size 30 (0x1e)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Private.CoreLib]System.Object::.ctor()
        IL_0006: nop
        IL_0007: ldarg.0
        IL_0008: ldarg.1
        IL_0009: ldfld int32 C::'<x>k__BackingField'
        IL_000e: stfld int32 C::'<x>k__BackingField'
        IL_0013: ldsfld int32 C::y
        IL_0018: stsfld int32 C::y
        IL_001d: ret
    } // end of method C::.ctor
```

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMYHsEBNkA0MIAlgDYG4gDUAPrrFAAQACATEwMIAUJAdjCYAPAJQBYAFABvSUzmsAjADYm/QQE8mAXiYK2AZgDckgL6TJfKAFsUABygYIigHQAlOAJI3nnLFdvkEAgAykEAbiSOyJIyEvKs+kzI0GQQuKwcAJLIAKJCMEGWZJl8JDCy8rHxZhI1QA==

### PR DESCRIPTION
**Pull request by AlekseyTs**
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-01.md
Also fixes #44879.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@agocke, @RikkiGibson, @cston, @jcouv, @dotnet/roslyn-compiler Please review, need a second sign-off, the amount of changes is small.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LookupPositionTests.cs [hunk 1]
```diff
@@ -2043,6 +2043,81 @@ interface C : Base(X)
...
+                    "System.Object System.Object.MemberwiseClone()",
+                    "System.String System.Object.ToString()",
+                    "System.Type C.EqualityContract { get; }",
+                    "System.Type System.Object.GetType()",
+                    "void System.Object.Finalize()"),
+                s_pop
+            );
+
+            TestLookupNames(text, expectedNames);
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LookupPositionTests.cs:2119]: Consider testing a `const` field as well.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LookupPositionTests.cs:2119]: > Consider testing a const field as well.

Const fields are static fields, so I am pretty confident this test is fairly equivalent to verify scoping rules. I will add a test like that for completeness sake in a separate PR 

---
In reply to: [436281327](https://github.com/dotnet/roslyn/pull/44906#discussion_r436281327) [](ancestors = 436281327)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 2]
```diff
@@ -6081,5 +6090,149 @@ static void Main()
...
+            Assert.Equal("= X + 1", x.Parent!.Parent!.ToString());
+
+            var symbol = model.GetSymbolInfo(x).Symbol;
+            Assert.Equal(SymbolKind.Property, symbol!.Kind);
+            Assert.Equal("System.Int32 C.X { get; init; }", symbol.ToTestDisplayString());
+            Assert.Equal("C", symbol.ContainingSymbol.ToTestDisplayString());
+            Assert.Equal("System.Int32 C.Z", model.GetEnclosingSymbol(x.SpanStart).ToTestDisplayString());
+            Assert.Contains(symbol, model.LookupSymbols(x.SpanStart, name: "X"));
+            Assert.Contains("X", model.LookupNames(x.SpanStart));
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6236]: Consider testing capturing:
```C#
record C(int X)
{
    Func<int> F = () => X;
}
```

**Comment by jcouv** [comment]
FYI, resolved conflicts following Chuck's PR.

**Comment by RikkiGibson** [comment]
@jcouv there are analyzer diagnostics in the Correctness run.

**Pull request merged**